### ISSUE TITLE
Questions: Categorical

### ISSUE THREAD

**Issue by AllardJM** [dswah/pyGAM#131]
Great lib! I saw some nice results on a model I built (regression). I am curious though:

1) I see warnings about features being categorical when there are none - how is this determined and what is the impact?

2) If a new observation has a level outside the training set, an error occurs. One way I *THINK* to handle this and make sure scoring doesnt fail, is to clip all predictors to the min and max observed when training the model - is this correct? 

3) Related to #2 is there a way to export the expected ranges from the model object or best to just use min and max of training records for each predictor?

**Comment by dswah** [comment]
thanks @AllardJM!

ah, yes, the automatic data detection sometimes is not very useful :)

1) categorical data types must satisfy 2 criteria (link to code [1])
- number of unique values should be below a certain ratio. intuition is that numerical data should not coincide very frequently, while this is certainly possible for categorical types.
- the data is encoded such that the min is 0, and the max is the number of unique values minus 1.

the impact is that if you see a warning about categorical data, when there is none, then your model is not being smoothed as you would expect. numerical data have smoothing among the coefficients of consecutive splines which forces them to be more similar and the overall function to be smoother. for categorical data, there is no such smoothing, only an l2 penalty on coefficient size.

to avoid this problem, you can manually set the data types. when you start your model do:

```
gam = LinearGAM(dtypes=['auto', 'auto', 'numerical', 'auto'...]) 
```
this will force the 3rd feature to be numerical, and allow the others to be selected automatically.

2) correct: for categorical data there will be an error whenever a new observation contains an unseen category. i dont think clipping an unseen category to the training range is correct, because often times categories are not ordinal at all. imagine is your categories are [dog, cat] and then in testing you see [bird]. it wouldnt seem right to set [bird] -> [cat]. I followed sklearn convention for their `LabelEncoder` where the encoder throws an error when it encounters an unseen category.

This error doesnt happen for numerical data, where unseen values are simply extrapolated linearly, or interpolated. 

3) yes, you can see the ranges of each feature. take your fitted GAM and check:
`gam._edge_knots`


[1]
Data type detection code
(https://github.com/dswah/pyGAM/blob/master/pygam/utils.py#L140)

**Comment by AllardJM** [comment]
Really helpful, thanks! Clipping seems like an OK approach to deal with a model I built where there were some  variables that were unary in training so were treated categorical, but then a test set had another value. These variables were numeric but should have really been removed before the model was built.

**Comment by AllardJM** [comment]
BTW you may be interested to know, the regression data I had was small, about 5,000 rows and 100 columns but pygam produced a better model on the test set than a carefully tuned xgboost :)

**Comment by dswah** [comment]
awesome :)
